mlflow/models/utils.py

```diff
@@ -270,89 +283,113 @@ def __init__(self, input_example: ModelInputExample):
+                            "type": "ndarray",
+                            "format": "tf-serving",
+                        }
+                    )
+                else:
+                    self._inference_data = pd.DataFrame(input_example)
+                    self.data = _convert_dataframe_to_split_dict(self._inference_data)
+                    self.serving_input = {DF_SPLIT: self.data}
+                    self.info.update(
+                        {
+                            "type": "dataframe",
+                            "pandas_orient": "split",
+                        }
+                    )
             else:
-                self._inference_data = pd.DataFrame(input_example)
-                self.data = _handle_dataframe_input(self._inference_data)
+                self._inference_data = _coerce_to_pandas_df(input_example)
+                if self._inference_data is None:
+                    raise TypeError(
+                        "Expected one of the following types:\n"
+                        "- pandas.DataFrame\n"
+                        "- numpy.ndarray\n"
+                        "- dictionary of (name -> numpy.ndarray)\n"
+                        "- scipy.sparse.csr_matrix\n"
+                        "- scipy.sparse.csc_matrix\n"
+                        "- dict\n"
+                        "- list\n"
+                        "- scalars\n"
+                        f"but got '{type(input_example)}'",
+                    )
+                self.data = _convert_dataframe_to_split_dict(self._inference_data)
+                self.serving_input = {DF_SPLIT: self.data}
+                orient = "split" if "columns" in self.data else "values"
                 self.info.update(
                     {
                         "type": "dataframe",
-                        "pandas_orient": "split",
+                        "pandas_orient": orient,
                     }
                 )
-        else:
-            self._inference_data = _coerce_to_pandas_df(input_example)
-            if self._inference_data is None:
-                raise TypeError(
-                    "Expected one of the following types:\n"
-                    "- pandas.DataFrame\n"
-                    "- numpy.ndarray\n"
-                    "- dictionary of (name -> numpy.ndarray)\n"
-                    "- scipy.sparse.csr_matrix\n"
-                    "- scipy.sparse.csc_matrix\n"
-                    "- dict\n"
-                    "- list\n"
-                    "- scalars\n"
-                    f"but got '{type(input_example)}'",
-                )
-            self.data = _handle_dataframe_input(self._inference_data)
-            orient = "split" if "columns" in self.data else "values"
-            self.info.update(
-                {
-                    "type": "dataframe",
-                    "pandas_orient": orient,
-                }
-            )
 
+    # TODO: isolate to make sure `artifact_path` saves an example that can be
```

Comment:
Currently self.data is a bit confusing as it sometimes has the `inputs` key (similar to serving input) but sometimes it doesn't (for dataframe input). My thoughts on this is to clean up it in a follow-up PR, after we switch to read example from serving_input_path on UI, we should clean up self.data and only store the data that pyfunc_model.predict can inference directly.

Reply:
This makes sense :)